feat: replace pyjwkest with pyjwt

mumarkhan999 opened the pull request:
Things to be done in this PR

This is part of https://github.com/openedx/edx-drf-extensions/issues/290

- [X] Upgraded JWT_SIGNING_KEYS using https://github.com/openedx/edx-platform/blob/master/scripts/jwk-precompute-params.py  
- [X] Replaced Pyjwkest usage with Pyjwt in codebase
- [X] Docs updated
- [X] Tests updated 
- [x] edx-drf-extesion is being installed from pypi
- [x] https://github.com/openedx/edx-drf-extensions/pull/354 is merged

timmc-edx commented:
The source changes so far look good. There are still some references to pyjwkest in the repo, though: Tests, a management command, docs, example code in an ADR, and the requirements files. The PR title makes me think the intent is to remove it entirely. If not, what boundary are you hoping to draw for this PR?

mumarkhan999 commented:
> The source changes so far look good. There are still some references to pyjwkest in the repo, though: Tests, a management command, docs, example code in an ADR, and the requirements files. The PR title makes me think the intent is to remove it entirely. If not, what boundary are you hoping to draw for this PR?

Except for management command and requirements files, all the above-mentioned tasks will be done in this PR.

timmc-edx commented on openedx/core/djangoapps/oauth_dispatch/tests/mixins.py:
```diff
@@ -33,25 +33,37 @@ def _decode_jwt(verify_expiration):
             verify the expiration of said token
             """
-            keys = KEYS()
+            key_set = []
```
It looks like this code largely duplicates what's in edx-drf-extensions. Can we call that code instead?

(Having the duplicated code increases maintenance burden but also risks us having a mismatch in behavior between the tests and how we actually verify and decode JWTs.)

mumarkhan999 replied:
I'm planning to move JWT-related similar code to this repo https://github.com/edx/token-utils.  But that repo is under `edx` organization. What's your say about that?

robrap replied:
It's possible that move makes sense, but it deserves as ADR, and will have some consequences:
- Ownership of token-utils is currently owned by cosmonauts, and I imagine we would need a discussion about that. Would it move to arbi-bom or arch-bom? We'd need to discuss with both teams.
- Yes, if this became a required dependency, it would need to move to the `openedx` repo.
- Would these utilities be exposed via edx-drf-extensions as well, or just through the new repo? Would it depend on whether or not it was specifically JWT auth related?
- The repo itself is not in a great state. It looks like https://github.com/edx/token-utils/blob/main/docs/decisions/0001-purpose-of-this-repo.rst was never updated, and I imagine other docs, so its purpose is not yet clearly documented.

Things to consider.

timmc-edx replied:
It's hard to say whether a move to token-utils is a good idea without knowing more about what that repo is supposed to be. (It's not clear from the README or other docs.) I am interested in knowing more about it, but it feels out of scope for the present work.

Here's what I was picturing for deduplication: https://github.com/openedx/edx-platform/pull/32466 -- this rewrites the mixin to use Django setting overrides so that it can just call our standard decoders.

robrap replied:
> It's hard to say whether a move to token-utils is a good idea without knowing more about what that repo is supposed to be. (It's not clear from the README or other docs.)

The repo token-utils was created for edx-exams, where they wanted to create custom signed JWTs (i.e. non-auth JWTs). I'll leave it to you all decide how to proceed, but it is important to note that edx-exams has its own token signing which will need to go through this same process. See https://github.com/edx/edx-internal/blob/2785fc7599e1f39fbd5387c3da18c0e74883f4b2/argocd/applications/edx-exams/prod-config.yaml#L102-L104 for the config. Note: it seems this setting was designed to just allow a single signing key per service for **all** custom signed JWTs. Not sure if that is the best design, but that is certainly not an issue for today.

robrap replied:
@mumarkhan999: When you get the chance, can you confirm that you read this and you are aware of the other `JWT_PRIVATE_SIGNING_JWK` cases? This is all I see using https://github.com/search?q=org%3Aedx%20JWT_PRIVATE_SIGNING_JWK&type=code, expect a local config in edx-exams.

mumarkhan999 replied:
@robrap Yeah I read it. And No I wasn't aware of this local config.
Will look into that once start working on that repo.

robrap replied:
@mumarkhan999: Is there a top-level Github Issue or Confluence page tracking this work across all services where we can document these types of things?

mumarkhan999 replied:
@robrap Yes we have one. Here's the link
https://github.com/openedx/edx-drf-extensions/issues/290

The ticket does not have much detail for now. But planning to add some there.

robrap replied:
Thanks. I added a link to that ticket in the PR description. I think it would be useful to add references to the top-level ticket in PR descriptions and/or commit comments so we could find all the parts from the top-level issue as needed.

timmc-edx commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0008-use-asymmetric-jwts.rst:
```diff
@@ -89,62 +89,78 @@ KeyPair Generation
 
 Here is code for generating a keypair::
 
+    import json
```
I wonder if rather than having all of this example code to maintain (and that isn't covered by tests) we should just point to the places we actually use the code. There's a management command for generating and serializing keypairs, and there's code in edx-platform and edx-drf-extensions for signing and validating.

My only concerns with that would be 1) code can move over time, so maybe we'd want to use permalinks to code at specific commits, and 2) some of the code is overly complicated because it still has branches for symmetric signing.

@robrap Any thoughts on this? It's the second time recently I've run into example code in ADRs.

robrap replied:
I agree that ADRs might include pseudo-code to add clarity, but shouldn't have actual code. I am only responding to @timmc-edx's comment, and I am not looking at this code.

mumarkhan999 commented on requirements/edx/base.txt:
```diff
@@ -471,9 +471,9 @@ edx-django-utils==5.4.0
     #   outcome-surveys
     #   super-csv
-edx-drf-extensions==8.8.0
+edx-drf-extensions @ git+https://github.com/openedx/edx-drf-extensions.git@umar/make-verify-signature-method-generic
```
📢📢📢
🛠️🛠️🛠️
This change is just for testing.
And only related to this PR.
Will be revered in this PR before merging.

mumarkhan999 commented on openedx/core/djangoapps/oauth_dispatch/tests/mixins.py:
```diff
@@ -33,25 +31,14 @@ def _decode_jwt(verify_expiration):
-                    "verify_exp": verify_expiration
-                },
-            )
+            key_set = get_signing_jwk_key_set(secret_key, add_asymmetric_keys=should_be_asymmetric_key)
```
@timmc-edx 
Now we're using the same functionality that we implemented in https://github.com/openedx/edx-drf-extensions/pull/354/files.

We just needed to make that functionality a bit generic.

mumarkhan999 committed:
chore: replace pyjwkest with pyjwt

mumarkhan999 committed:
squash! update ADR

mumarkhan999 committed:
squash! update ADR

mumarkhan999 committed:
squash! update ADR

mumarkhan999 committed:
squash! update ADR

mumarkhan999 commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0008-use-asymmetric-jwts.rst:
```diff
@@ -47,23 +47,10 @@ be the only service configured with the aymmetric keypair, including its Private
 All other OAuth Clients will be configured with only the Public key portion of the asymmetric key pair.
 This will ultimately replace all uses of "symmetric" keys for signing JWTs.
 
-"kid" Key Identifier
```
Removing this block because of
```
Considering that we are doing verification
by looping through all the available public keys, the ``kid`` parameter is not
as important as it was before. But it is still a good practice to have it.
```

timmc-edx replied:
We should still recommend the use of a `kid` because it's useful in identifying keys in config files, and because it's very useful for adding observability -- during a key rotation, you can potentially watch the count of verifications using the first key ID drop to zero and the second key ID rise to replace them. (I seem to recall we added code to set a custom attribute based on `kid` but now I'm not sure if we actually did that or just planned to do it.)

mumarkhan999 replied:
Yeah, got your point.
I've brought back the old header in the doc. 
And updated this part as well
https://github.com/openedx/edx-platform/pull/32270/files#diff-5f5c0ae6e79da08673e8a1bf4ad3d8cdea340124f0cc6d6b77ca3493cce00199R154

mumarkhan999 commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0008-use-asymmetric-jwts.rst:
```diff
@@ -81,70 +68,56 @@ issuer, but with (the potential of) multiple signing keys stored in a JWT Set.
 
-Example Code
-------------
+Features
```
Instead of `explicitly` giving coding examples, just providing the links to the relevant methods/modules.

mumarkhan999 commented on openedx/core/djangoapps/oauth_dispatch/tests/test_views.py:
```diff
@@ -163,20 +161,6 @@ def _post_body(self, user, client, token_type=None, scope=None, asymmetric_jwt=N
 
         return body
 
-    def _generate_key_pair(self):
```
Removing this, as it is not in use.

timmc-edx commented on openedx/core/djangoapps/oauth_dispatch/docs/decisions/0008-use-asymmetric-jwts.rst:
```diff
@@ -81,70 +68,56 @@ issuer, but with (the potential of) multiple signing keys stored in a JWT Set.
+            "dp": "palm",
+            "dq": "pineapple",
+            "qi": "watermelon",
+            "kid": "your_key_id"
```
I think this would also be a good opportunity to switch away from the placeholder words to something that is more visually obvious as a placeholder:

```suggestion
                    "kty": "RSA",
                    "key_ops": ["verify"],
                    "n": "...",
                    "e": "...",
                    "kid": "your_key_id"
                }
            ]
        }
    """


    ## Private key
    """
        {
            "kty": "RSA",
            "key_ops": ["sign"],
            "n": "...",
            "e": "...",
            "d": "...",
            "p": "...",
            "q": "...",
            "dp": "...",
            "dq": "...",
            "qi": "...",
            "kid": "your_key_id"
```

timmc-edx replied:
We should also replace all these tab characters with consistent space-based indentation.

timmc-edx commented:
Other than the doc tweaks I think this is ready to go.

mumarkhan999 committed:
squash! update keys template

Co-authored-by: Tim McCormack <tmccormack@edx.org>

mumarkhan999 committed:
squash! update ADR

mumarkhan999 commented:
Closing to rerun stuck jobs.

timmc-edx reviewed (approved):
Looks great, thank you!

mumarkhan999 commented:
@timmc-edx 

Hi tim, while testing this PR on `sandbox` I got this `error`

```
    jwt = _create_jwt(request, user, expires_in)
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/user_authn/cookies.py", line 319, in _create_jwt
    return create_jwt_from_token(access_token, DOTAdapter(), use_asymmetric_key=True)
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/oauth_dispatch/jwt.py", line 110, in create_jwt_from_token
    jwt_token_dict = create_jwt_token_dict(token_dict, oauth_adapter, use_asymmetric_key)
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/oauth_dispatch/jwt.py", line 83, in create_jwt_token_dict
    jwt_access_token = _create_jwt(
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/oauth_dispatch/jwt.py", line 196, in _create_jwt
    return _encode_and_sign(payload, use_asymmetric_key, secret)
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/oauth_dispatch/jwt.py", line 286, in _encode_and_sign
    jwk = PyJWK(key, algorithm)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/jwt/api_jwk.py", line 60, in __init__
    self.key = self.Algorithm.from_jwk(self._jwk_data)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/jwt/algorithms.py", line 433, in from_jwk
    raise InvalidKeyError(
jwt.exceptions.InvalidKeyError: RSA key must include all parameters if any are present besides d
```

As far as I remember we have updated the keys. But it seems we have missed a few environments.
Can you please mention for which environments we've updated the keys?
What are the remaining ones?

And also can you please update the `keys` in the remaining `envs` so we can `merge` this PR?


Thanks.

robrap commented:
> Hi tim, while testing this PR on `sandbox` I got this `error`

Does this mean that https://github.com/edx/sandbox-internal was missed? (What about https://github.com/edx/edge-internal/ ?) Or were the changes elsewhere?

timmc-edx commented:
From my notes in https://github.com/edx/edx-arch-experiments/issues/261 it sounds like sandboxes actually generate their keypair during provisioning, rather than being configured with a static value. We deleted the configured sandbox keys some months ago, so something else is going on here. (See that ticket for a list of environments that were updated.)

mumarkhan999 commented:
I tested this `branch` on an already available `sandbox`  (triggered on `master` branch). On master the [generate_jwt_signing_key](https://github.com/search?q=org%3Aopenedx%20generate_jwt_signing_key&type=code) is outdated, that's why the keys are outdated.

Now I triggered a new one using the `current branch`. Let's see how it works.

kdmccormick commented:
Hey @mumarkhan999 , thank you for this upgrade. In the future, when you make a breaking change, please use an exclamation point in your commit message prefix, for example `chore!: replace pyjwkest with pyjwt`. That'll help Tutor maintainers notice the change and make the appropriate Tutor update.
